### PR TITLE
Refactor training loop with batch getter

### PR DESCRIPTION
**Pull request by billyang98**
cleaned up the training loop batches to look more like the facebook repo

Also we should do one step of each of the 4 tasks. This is more closely in line with both our papers, the Unsupervised MT and the TTS/ASR.

How I defined the number of epoch steps is based on one pass through for the discriminator's training set. When training one epoch of the discriminator, I think it is good practice to at least train over the whole dataset, or else by random chance, we may miss some examples. The disciminator's training set is also the largest, so this is not a problem for the other tasks.  When training without the discriminator, this means the model may go over the dataset some extra times, but I think this is worth it so that the number of training steps is equal for both training with and without the discriminator.

### CODE REVIEW

**Comment by billyang98** [comment]
Also note that this is on top of Lucas' branch. I will rebase any changes to that branch before merging

**Review by Lucaskabela — APPROVED** [review]
Looks great! Thanks for taking the initiative to clean this up even more

File: src/train.py [hunk 1]
```diff
@@ -17,9 +17,66 @@
...
+        self.unsupervised_iter = iter(DataLoader(unsupervised_dataset,
+            batch_size=self.batch_size, shuffle=True,
+            collate_fn=collate_fn_transformer, drop_last=True,
+            num_workers=self.num_workers))
+
+        self.discriminator_dataset = full_dataset
+        self.discriminator_iter = iter(DataLoader(full_dataset,
+            batch_size=self.batch_size, shuffle=True,
+            collate_fn=collate_fn_transformer, drop_last=True,
+            num_workers=self.num_workers))
```
Comment (liaojh1998) [inline src/train.py]: Instead of creating an `iter()` on a new `DataLoader()` every time, let's create the `DataLoader()` once, and call `iter()` on it every `StopIteration`. This will also avoid reshuffling the training data to keep training consistent and avoid repeat calling `collate_fn` on the big dataset (which is slow). Code would be like:

```python3
    def __init__(self, args, supervised_dataset, unsupervised_dataset, full_dataset):
        ...
        self.discriminator_dataloader = DataLoader(full_dataset,
            batch_size=self.batch_size, shuffle=True,
            collate_fn=collate_fn_transformer, drop_last=True,
            num_workers=self.num_workers)
        self.discriminator_iter = iter(self.discriminator_dataloader)
        ...

    def get_discriminator_batch(self):
        ...
        except StopIteration:
            self.discriminator_iter = iter(self.discriminator_dataloader)
            batch = self.discriminator_iter.next()  # or next(self.discriminator_iter)
        return batch
```

File: src/train.py [hunk 2]
```diff
@@ -199,42 +256,41 @@ def train(args):
...
+
+            # CM TTS/ASR
+            for _ in range(0, args.cm_steps):
+                batch = batch_getter.get_unsupervised_batch()
+                train_cm_step(losses, model, optimizer, batch, args)
+
+            # SUPERVISED
+            for _ in range(0, args.sp_steps):
+                batch = batch_getter.get_supervised_batch()
+                train_sp_step(losses, model, optimizer, batch, args)
```
Comment (liaojh1998) [inline src/train.py:278]: Nice and clean!

**Review by liaojh1998 — CHANGES_REQUESTED** [review]
Nice work, I have a comment on avoiding CPU usage during the creation of the `DataLoader()` from using `collate_fn` and shuffling.

**Review by liaojh1998 — APPROVED** [review]
Let’s get it
